core/server/src/main/java/alluxio/master/block/BlockMaster.java

```diff
@@ -77,21 +75,30 @@
 public final class BlockMaster extends AbstractMaster implements ContainerIdGenerable {
   private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
 
-  // Block metadata management.
   /**
-   * Blocks on all workers, including active and lost blocks. This state must be journaled. Access
-   * must be synchronized on mBlocks. If both block and worker metadata must be locked, mBlocks must
-   * be locked first.
+   * Concurrency and locking in the BlockMaster
+   *
+   * The block master uses concurrent data structures, to allow non-conflicting concurrent access.
+   * This means each piece of metadata should be locked individually. There are two types of
+   * metadata in the {@link BlockMaster}; {@link MasterBlockInfo} and {@link MasterWorkerInfo}.
+   * Individual objects must be locked before modifying the object, or reading a modifiable field
+   * of an object. This will protect the internal integrity of the metadata object.
+   *
+   * Lock ordering must be preserved in order to prevent deadlock. If both a worker and block
+   * metadata must be locked at the same time, the worker metadata ({@link MasterWorkerInfo})
```

Comment:
Is it possible to detect/enforce such locking order? For example, http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/util/concurrent/CycleDetectingLockFactory.html